### PR TITLE
Add hinge loss

### PR DESCRIPTION
**Pull request by mitmul**
This PR adds hinge loss. It can be used as a loss function for SVM. Here is an example: [chainer-svm](https://github.com/mitmul/chainer-svm).

### CODE REVIEW

File: chainer/functions/loss/hinge.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            ''', 'hinge_bwd')(self.xt, t, gloss, 1.0 / t.shape[0])
+
+        return gx, None
+
+
+def hinge(x, t, use_cudnn=True):
+    """Computes hinge loss as below:
+
+        .. math::
+            L = \sum_{n=1}^N max(0, 1 - x_n t_n)
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: Use `\\max`

File: chainer/functions/loss/hinge.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        where :math:`N` denotes the batchsize.
+
+    Args:
+        x (~chainer.Variable): Input variable. The shape of ``x`` should be the
+            same as ``t``
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: Please insert a period at the end of the line.

File: chainer/functions/loss/hinge.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+class Hinge(function.Function):
+
+    """Hinge loss."""
+
+    def __init__(self, use_cudnn=True):
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: There are no CuDNN version of hinge function.  `use_cudnn` argument is not required.

File: chainer/functions/loss/hinge.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            'T xt, S t, raw T gloss, T inv_cnt', 'T gx',
+            '''
+            if (xt < 1) gx = -gloss[0] * inv_cnt * t;
+            else gx = 0
+            ''', 'hinge_bwd')(self.xt, t, gloss, 1.0 / t.shape[0])
+
+        return gx, None
+
+
+def hinge(x, t, use_cudnn=True):
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: `use_cudnn` is not required.

**Comment by unnonouno** [comment]
I want to discuss about interface. I think `t \in {+1, -1}` is assumed, and multi-class classification is not supported. It is a little confusing, isn't it?
How about restring `t` to a boolean array, or accepting only one parameter `xt`?

**Comment by mitmul** [comment]
Yes, that's right. `t` accepts only +1 or -1 as its value. I think that letting `t` be a boolean array is better.

File: chainer/functions/loss/hinge.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            ''', 'hinge_bwd')(self.xt, t, gloss, 1.0 / t.shape[0])
+
+        return gx, None
+
+
+def hinge(x, t):
+    """Computes hinge loss as below:
+
+        .. math::
+            L = \\frac{1}{N} \sum_{n=1}^N \\max(0, 1 - x_n t_n)
```
Comment (okuta) [inline chainer/functions/loss/hinge.py]: Need two backslash. `\sum` -> `\\sum`.

**Comment by mitmul** [comment]
Updated to support multi-class classification task.

File: chainer/functions/loss/hinge.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+from chainer import function
+from chainer.utils import type_check
+
+
+class Hinge(function.Function):
+
+    """Hinge loss."""
+
+    def __init__(self, norm='L1'):
+        self.norm = norm
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: Can you check `norm in {'L1', 'L2'}`?

File: chainer/functions/loss/hinge.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    def forward_cpu(self, inputs):
+        x, t = inputs
+        num, dim = x.shape
+        self.bottom_diff = numpy.copy(x)
+        for i in six.moves.range(num):
+            self.bottom_diff[i, t[i]] *= -1
+        for i in six.moves.range(num):
+            for j in six.moves.range(dim):
+                self.bottom_diff[i, j] = max(0, 1 + self.bottom_diff[i, j])
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: `self.bottom_diff = numpy.max(0, 1 + self.bottom_diff)` ?

File: chainer/functions/loss/hinge.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        else:
+            raise NotImplementedError()
+
+        return cuda.cupy.array(loss, dtype=cuda.cupy.float32),
+
+    def backward_cpu(self, inputs, grad_outputs):
+        t, gloss = inputs[1], grad_outputs[0]
+        for i in six.moves.range(t.shape[0]):
+            self.bottom_diff[i, t[i]] *= -1
+        gx = numpy.empty_like(self.bottom_diff)
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: This line isn't required.

File: chainer/functions/loss/hinge.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        return gx, None
+
+    def backward_gpu(self, inputs, grad_outputs):
+        xp = cuda.get_array_module(*inputs)
+        t, gloss = inputs[1], grad_outputs[0]
+        self.bottom_diff = cuda.elementwise(
+            'S t, int32 dim', 'raw T bottom_diff',
+            'bottom_diff[i * dim + t] *= -1',
+            'hinge_bwd')(t, inputs[0].shape[1], self.bottom_diff)
+        gx = xp.empty_like(self.bottom_diff)
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: This line isn't required.

File: chainer/functions/loss/hinge.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        if self.norm == 'L1':
+            gx = gloss * xp.sign(self.bottom_diff) / t.shape[0]
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: `gx = (gloss / t.shape[0]) * xp.sign(self.bottom_diff.)` is faster maybe.

File: chainer/functions/loss/hinge.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        elif self.norm == 'L2':
+            gx = gloss * 2 * self.bottom_diff / t.shape[0]
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: `(gloss * 2 / t.shape[0]) * self.bottom_diff` is faster maybe.

File: chainer/functions/loss/hinge.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        num, dim = x.shape
+        self.bottom_diff = cuda.cupy.copy(x)
+        self.bottom_diff = cuda.elementwise(
+            'S t, int32 dim', 'raw T bottom_diff',
+            'bottom_diff[i * dim + t] *= -1',
+            'hinge_fwd')(t, dim, self.bottom_diff)
+        self.bottom_diff = cuda.cupy.maximum(0, 1 + self.bottom_diff)
+        loss = 0
+        if self.norm == 'L1':
+            loss = cuda.cupy.sum(cuda.cupy.abs(self.bottom_diff)) / num
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: Is `abs` required? Each element in `bottom_diff` is positive, isn't it?

**Comment by mitmul** [comment]
@unnonouno Thank you for your precise review. I think all points you pointed out has been fixed. Please check them again, thank you.

File: chainer/functions/loss/hinge.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+        return gx, None
+
+
+def hinge(x, t, norm='L1'):
+    """Computes the hinge loss for a one-of-many classification task.
+
+        .. math::
+            L = \\frac{1}{N} \\sum_{n=1}^N \\sum_{k=1}^K \\left[
+            \\max(0, 1 - \\delta\\{l_n = k\\} t_{nk}) \\right]^p
```
Comment (delta2323) [inline chainer/functions/loss/hinge.py]: Is it intentional that `{l_n = k}` is not lowered.

File: chainer/functions/loss/hinge.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+        where :math:`N` denotes the batchsize, :math:`K` is the number of
+        classes of interest,
+
+        .. math::
+            \\delta \\{ {\\rm condition} \\} = \\left \\{ \\begin{array}{cc}
+            1 & {\\rm if~condition} \\\\
+            -1 & {\\rm otherwise,}
+            \\end{array} \\right.
+
```
Comment (delta2323) [inline chainer/functions/loss/hinge.py]: It is well-mannered to add **and** here.

File: chainer/functions/loss/hinge.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+            p = \\left \\{ \\begin{array}{cc}
+            1 & {\\rm if~norm} = {\\rm 'L1'} \\\\
+            2 & {\\rm if~norm} = {\\rm 'L2'.}
+            \\end{array} \\right.
+
+    Args:
+        x (~chainer.Variable): Input variable. The shape of ``x`` should be
+            (:math:`N`, :math:`K`).
+        t (~chainer.Variable): The :math:`N`-dimensional label vector
+            :math:`{\\bf l}` with values :math:`l_n \in [0, 1, 2, \dots, K-1]`.
```
Comment (delta2323) [inline chainer/functions/loss/hinge.py]: Use curly parentheses to indicate a set (`{0, 1, \dots, K-1}`)

File: chainer/functions/loss/hinge.py (hunk 13 above)
Comment (mitmul) [inline chainer/functions/loss/hinge.py]: I followed [Caffe doc](http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1HingeLossLayer.html). Should that be a subscript?

**Comment by unnonouno** [comment]
It looks good to me. Could you merge or rebase the master?

File: chainer/functions/loss/hinge.py (hunk 13 above)
Comment (delta2323) [inline chainer/functions/loss/hinge.py]: Then, I think it is OK to leave it as it is.

File: chainer/functions/loss/hinge.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    def forward_cpu(self, inputs):
+        x, t = inputs
+        num, dim = x.shape
+        self.bottom_diff = numpy.copy(x)
+        for i in six.moves.range(num):
+            self.bottom_diff[i, t[i]] *= -1
+        self.bottom_diff = numpy.maximum(0, 1 + self.bottom_diff)
+        loss = 0
+        if self.norm == 'L1':
+            loss = numpy.sum(numpy.abs(self.bottom_diff)) / num
```
Comment (unnonouno) [inline chainer/functions/loss/hinge.py]: Sorry, this `abs` isn't necessary.

**Comment by unnonouno** [comment]
Thank you very much! Great work!!
